File: OpenBurnBarCore/Sources/OpenBurnBarSignalCore/CloudVaultDeviceTrustChain.swift
import Foundation
import LibSignalClient

public struct CloudVaultDeviceTrustChainPayload: Sendable, Hashable {
    public let uid: String
    public let targetDeviceId: String
    public let targetEscrowPublicKeyFingerprint: String
    public let targetKeyVersion: Int
    public let targetSignalIdentityKeyId: String
    public let targetSignalIdentityPublicKeyFingerprint: String
    public let approverDeviceId: String
    public let approverSignalIdentityKeyId: String
    public let approverSignalIdentityPublicKeyFingerprint: String

    public init(
        uid: String,
        targetDeviceId: String,
        targetEscrowPublicKeyFingerprint: String,
        targetKeyVersion: Int,
        targetSignalIdentityKeyId: String,
        targetSignalIdentityPublicKeyFingerprint: String,
        approverDeviceId: String,
        approverSignalIdentityKeyId: String,
        approverSignalIdentityPublicKeyFingerprint: String
    ) {
        self.uid = uid
        self.targetDeviceId = targetDeviceId
        self.targetEscrowPublicKeyFingerprint = targetEscrowPublicKeyFingerprint
        self.targetKeyVersion = targetKeyVersion
        self.targetSignalIdentityKeyId = targetSignalIdentityKeyId
        self.targetSignalIdentityPublicKeyFingerprint = targetSignalIdentityPublicKeyFingerprint
        self.approverDeviceId = approverDeviceId
        self.approverSignalIdentityKeyId = approverSignalIdentityKeyId
        self.approverSignalIdentityPublicKeyFingerprint = approverSignalIdentityPublicKeyFingerprint
    }
}

public enum CloudVaultDeviceTrustChain {
    public static let version = 1
    public static let algorithm = "signal-identity-xeddsa-v1"
    public static let domain = "OpenBurnBar-CloudVault-DeviceTrust-v1"

    public static func canonicalPayload(_ payload: CloudVaultDeviceTrustChainPayload) -> Data {
        let segments = [
            "uid", payload.uid,
            "targetDeviceId", payload.targetDeviceId,
            "targetEscrowPublicKeyFingerprint", payload.targetEscrowPublicKeyFingerprint,
            "targetKeyVersion", "\(payload.targetKeyVersion)",
            "targetSignalIdentityKeyId", payload.targetSignalIdentityKeyId,
            "targetSignalIdentityPublicKeyFingerprint", payload.targetSignalIdentityPublicKeyFingerprint,
            "approverDeviceId", payload.approverDeviceId,
            "approverSignalIdentityKeyId", payload.approverSignalIdentityKeyId,
            "approverSignalIdentityPublicKeyFingerprint", payload.approverSignalIdentityPublicKeyFingerprint
        ]
        var canonical = "\(domain)\n"
        for segment in segments {
            let utf8Count = segment.data(using: .utf8)?.count ?? 0
            canonical += "\(utf8Count):\(segment)\n"
        }
        return Data(canonical.utf8)
    }

    public static func sign(
        _ payload: CloudVaultDeviceTrustChainPayload,
        approverIdentity: OpenBurnBarSignalIdentityKeypair
    ) throws -> String {
        let privateKey = try PrivateKey(approverIdentity.privateKeyData)
        return privateKey.generateSignature(message: canonicalPayload(payload)).base64EncodedString()
    }

    public static func verify(
        _ payload: CloudVaultDeviceTrustChainPayload,
        signatureBase64: String,
        approverPublicKeyData: Data
    ) -> Bool {
        guard let signature = Data(base64Encoded: signatureBase64) else {
            return false
        }
        do {
            let publicKey = try PublicKey(approverPublicKeyData)
            return try publicKey.verifySignature(message: canonicalPayload(payload), signature: signature)
        } catch {
            return false
        }
    }
}

public struct CloudVaultTrustedDeviceActionProofPayload: Sendable, Hashable {
    public let uid: String
    public let deviceId: String
    public let actionKind: String
    public let subjectId: String
    public let approve: Bool
    public let nonce: String
    public let issuedAtMillis: Int64
    public let deviceSignalIdentityKeyId: String
    public let deviceSignalIdentityPublicKeyFingerprint: String

    public init(
        uid: String,
        deviceId: String,
        actionKind: String,
        subjectId: String,
        approve: Bool,
        nonce: String,
        issuedAtMillis: Int64,
        deviceSignalIdentityKeyId: String,
        deviceSignalIdentityPublicKeyFingerprint: String
    ) {
        self.uid = uid
        self.deviceId = deviceId
        self.actionKind = actionKind
        self.subjectId = subjectId
        self.approve = approve
        self.nonce = nonce
        self.issuedAtMillis = issuedAtMillis
        self.deviceSignalIdentityKeyId = deviceSignalIdentityKeyId
        self.deviceSignalIdentityPublicKeyFingerprint = deviceSignalIdentityPublicKeyFingerprint
    }
}

public enum CloudVaultTrustedDeviceActionProof {
    public static let version = 1
    public static let algorithm = CloudVaultDeviceTrustChain.algorithm
    public static let domain = "OpenBurnBar-TrustedDeviceAction-v1"

    public static func canonicalPayload(_ payload: CloudVaultTrustedDeviceActionProofPayload) -> Data {
        let segments = [
            "uid", payload.uid,
            "deviceId", payload.deviceId,
            "actionKind", payload.actionKind,
            "subjectId", payload.subjectId,
            "approve", payload.approve ? "true" : "false",
            "nonce", payload.nonce,
            "issuedAtMillis", "\(payload.issuedAtMillis)",
            "deviceSignalIdentityKeyId", payload.deviceSignalIdentityKeyId,
            "deviceSignalIdentityPublicKeyFingerprint", payload.deviceSignalIdentityPublicKeyFingerprint
        ]
        var canonical = "\(domain)\n"
        for segment in segments {
            let utf8Count = segment.data(using: .utf8)?.count ?? 0
            canonical += "\(utf8Count):\(segment)\n"
        }
        return Data(canonical.utf8)
    }

    public static func sign(
        _ payload: CloudVaultTrustedDeviceActionProofPayload,
        identity: OpenBurnBarSignalIdentityKeypair
    ) throws -> String {
        let privateKey = try PrivateKey(identity.privateKeyData)
        return privateKey.generateSignature(message: canonicalPayload(payload)).base64EncodedString()
    }

    public static func verify(
        _ payload: CloudVaultTrustedDeviceActionProofPayload,
        signatureBase64: String,
        publicKeyData: Data
    ) -> Bool {
        guard let signature = Data(base64Encoded: signatureBase64) else {
            return false
        }
        do {
            let publicKey = try PublicKey(publicKeyData)
            return try publicKey.verifySignature(message: canonicalPayload(payload), signature: signature)
        } catch {
            return false
        }
    }
}

File: android/app/src/main/java/com/openburnbar/data/cloud/CloudVaultDeviceTrustChain.kt
package com.openburnbar.data.cloud

import org.signal.libsignal.protocol.ecc.ECPublicKey

data class CloudVaultDeviceTrustChainPayload(
    val uid: String,
    val targetDeviceId: String,
    val targetEscrowPublicKeyFingerprint: String,
    val targetKeyVersion: Int,
    val targetSignalIdentityKeyId: String,
    val targetSignalIdentityPublicKeyFingerprint: String,
    val approverDeviceId: String,
    val approverSignalIdentityKeyId: String,
    val approverSignalIdentityPublicKeyFingerprint: String,
)

data class CloudVaultDeviceTrustChainProof(
    val version: Int = CloudVaultDeviceTrustChain.VERSION,
    val algorithm: String = CloudVaultDeviceTrustChain.ALGORITHM,
    val targetSignalIdentityKeyId: String,
    val targetSignalIdentityPublicKeyFingerprint: String,
    val approverSignalIdentityKeyId: String,
    val approverSignalIdentityPublicKeyFingerprint: String,
    val signature: String,
) {
    fun asMap(): Map<String, Any> =
        mapOf(
            "version" to version,
            "algorithm" to algorithm,
            "targetSignalIdentityKeyId" to targetSignalIdentityKeyId,
            "targetSignalIdentityPublicKeyFingerprint" to targetSignalIdentityPublicKeyFingerprint,
            "approverSignalIdentityKeyId" to approverSignalIdentityKeyId,
            "approverSignalIdentityPublicKeyFingerprint" to approverSignalIdentityPublicKeyFingerprint,
            "signature" to signature,
        )
}

data class CloudVaultTrustedDeviceActionProofPayload(
    val uid: String,
    val deviceId: String,
    val actionKind: String,
    val subjectId: String,
    val approve: Boolean,
    val nonce: String,
    val issuedAtMillis: Long,
    val deviceSignalIdentityKeyId: String,
    val deviceSignalIdentityPublicKeyFingerprint: String,
)

data class CloudVaultTrustedDeviceActionProof(
    val version: Int = CloudVaultTrustedDeviceActionProofSigner.VERSION,
    val algorithm: String = CloudVaultTrustedDeviceActionProofSigner.ALGORITHM,
    val deviceSignalIdentityKeyId: String,
    val deviceSignalIdentityPublicKeyFingerprint: String,
    val issuedAtMillis: Long,
    val signature: String,
) {
    fun asMap(): Map<String, Any> =
        mapOf(
            "version" to version,
            "algorithm" to algorithm,
            "deviceSignalIdentityKeyId" to deviceSignalIdentityKeyId,
            "deviceSignalIdentityPublicKeyFingerprint" to deviceSignalIdentityPublicKeyFingerprint,
            "issuedAtMillis" to issuedAtMillis,
            "signature" to signature,
        )
}

object CloudVaultTrustedDeviceActionProofSigner {
    const val VERSION: Int = 1
    const val ALGORITHM: String = CloudVaultDeviceTrustChain.ALGORITHM
    private const val DOMAIN: String = "OpenBurnBar-TrustedDeviceAction-v1"

    fun canonicalPayload(payload: CloudVaultTrustedDeviceActionProofPayload): ByteArray {
        val segments =
            listOf(
                "uid",
                payload.uid,
                "deviceId",
                payload.deviceId,
                "actionKind",
                payload.actionKind,
                "subjectId",
                payload.subjectId,
                "approve",
                if (payload.approve) "true" else "false",
                "nonce",
                payload.nonce,
                "issuedAtMillis",
                payload.issuedAtMillis.toString(),
                "deviceSignalIdentityKeyId",
                payload.deviceSignalIdentityKeyId,
                "deviceSignalIdentityPublicKeyFingerprint",
                payload.deviceSignalIdentityPublicKeyFingerprint,
            )
        val builder = StringBuilder()
        builder.append(DOMAIN).append('\n')
        for (segment in segments) {
            builder.append(segment.toByteArray(Charsets.UTF_8).size)
                .append(':')
                .append(segment)
                .append('\n')
        }
        return builder.toString().toByteArray(Charsets.UTF_8)
    }

    fun sign(
        payload: CloudVaultTrustedDeviceActionProofPayload,
        identity: AndroidSignalIdentityKeypair,
    ): String {
        val signature =
            CloudVaultCryptoSupport.decodeSignalPrivateKey(identity.privateKeyData)
                .calculateSignature(canonicalPayload(payload))
        return CloudVaultCryptoSupport.encodeBase64(signature)
    }

    fun verify(
        payload: CloudVaultTrustedDeviceActionProofPayload,
        signatureBase64: String,
        publicKeyData: ByteArray,
    ): Boolean =
        runCatching {
            val signature = CloudVaultCryptoSupport.decodeBase64(signatureBase64)
            ECPublicKey(publicKeyData).verifySignature(canonicalPayload(payload), signature)
        }.getOrDefault(false)
}

object CloudVaultDeviceTrustChain {
    const val VERSION: Int = 1
    const val ALGORITHM: String = "signal-identity-xeddsa-v1"
    private const val DOMAIN: String = "OpenBurnBar-CloudVault-DeviceTrust-v1"

    fun canonicalPayload(payload: CloudVaultDeviceTrustChainPayload): ByteArray {
        val segments =
            listOf(
                "uid",
                payload.uid,
                "targetDeviceId",
                payload.targetDeviceId,
                "targetEscrowPublicKeyFingerprint",
                payload.targetEscrowPublicKeyFingerprint,
                "targetKeyVersion",
                payload.targetKeyVersion.toString(),
                "targetSignalIdentityKeyId",
                payload.targetSignalIdentityKeyId,
                "targetSignalIdentityPublicKeyFingerprint",
                payload.targetSignalIdentityPublicKeyFingerprint,
                "approverDeviceId",
                payload.approverDeviceId,
                "approverSignalIdentityKeyId",
                payload.approverSignalIdentityKeyId,
                "approverSignalIdentityPublicKeyFingerprint",
                payload.approverSignalIdentityPublicKeyFingerprint,
            )
        val builder = StringBuilder()
        builder.append(DOMAIN).append('\n')
        for (segment in segments) {
            builder.append(segment.toByteArray(Charsets.UTF_8).size)
                .append(':')
                .append(segment)
                .append('\n')
        }
        return builder.toString().toByteArray(Charsets.UTF_8)
    }

    fun sign(
        payload: CloudVaultDeviceTrustChainPayload,
        approverIdentity: AndroidSignalIdentityKeypair,
    ): String {
        val signature =
            CloudVaultCryptoSupport.decodeSignalPrivateKey(approverIdentity.privateKeyData)
                .calculateSignature(canonicalPayload(payload))
        return CloudVaultCryptoSupport.encodeBase64(signature)
    }

    fun verify(
        payload: CloudVaultDeviceTrustChainPayload,
        signatureBase64: String,
        approverPublicKeyData: ByteArray,
    ): Boolean =
        runCatching {
            val signature = CloudVaultCryptoSupport.decodeBase64(signatureBase64)
            ECPublicKey(approverPublicKeyData).verifySignature(canonicalPayload(payload), signature)
        }.getOrDefault(false)
}
